stac_pydantic/item.py

```diff
@@ -28,38 +25,19 @@ class ItemProperties(StacCommonMetadata):
     https://github.com/radiantearth/stac-spec/blob/v1.0.0/item-spec/item-spec.md#properties-object
     """
 
-    datetime: Union[dt, str] = Field(..., alias="datetime")
+    datetime: Optional[dt] = Field(..., alias="datetime")
 
     # Check https://docs.pydantic.dev/dev-v2/migration/#changes-to-config for more information.
     model_config = ConfigDict(extra="allow")
 
-    @model_validator(mode="before")
-    @classmethod
-    def validate_datetime(cls, data: Dict[str, Any]) -> Dict[str, Any]:
-        datetime = data.get("datetime")
-        start_datetime = data.get("start_datetime")
-        end_datetime = data.get("end_datetime")
-
-        if not datetime or datetime == "null":
-            if not start_datetime and not end_datetime:
-                raise ValueError(
-                    "start_datetime and end_datetime must be specified when datetime is null"
-                )
-
-        if isinstance(datetime, str):
-            data["datetime"] = parse_datetime(datetime)
-
-        if isinstance(start_datetime, str):
-            data["start_datetime"] = parse_datetime(start_datetime)
-
-        if isinstance(end_datetime, str):
-            data["end_datetime"] = parse_datetime(end_datetime)
-
-        return data
+    @model_validator(mode="after")
+    def validate_datetime(self) -> "ItemProperties":
+        if not self.datetime and (not self.start_datetime or not self.end_datetime):
+            raise ValueError(
+                "start_datetime and end_datetime must be specified when datetime is null"
+            )
 
-    @field_serializer("datetime")
-    def serialize_datetime(self, v: dt, _info: Any) -> str:
-        return v.strftime(DATETIME_RFC339)
```

Comment:
I think we still need the serializer to make sure we respect the RFC3339 spec

This could be one implementation (to make sure we keep microseconds, if provided) 

```python
    @field_serializer("datetime", "start_datetime", "end_datetime", when_used="always")
    def serialize_datetime(self, v: Optional[dt], _info: Any) -> str:
        if v:
            if str(v)[19] == ".":
                return v.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
            else:
                return v.strftime("%Y-%m-%dT%H:%M:%SZ")
```

Reply:
Assuming the datetime object has all the required values, then the pydantic output will always be valid RFC3339. But that is not always the case. Pydantic allows in time without seconds, but then the output would not be valid. If input is validated through `parse_rfc3339` then it would be fine to use no serializer and let pydantic handle it. 

If, however, we want to accept a wider range of inputs (`AwareDatetime`) then a serializer is probably necessary. I believe the best solution would actually be to do `.isoformat()` on the datetime object, not an `.strftime`. As RFC3339 allows any timezone. 
```
   time-numoffset  = ("+" / "-") time-hour ":" time-minute
   time-offset     = "Z" / time-numoffset
```
And the use of `.strftime("%Y-%m-%dT%H:%M:%S.%fZ")` is not guaranteed to be correct. Depending on the input timezone. 
```python
from ciso8601 import parse_rfc3339
from datetime import datetime, timezone
from pydantic import TypeAdapter
d = parse_rfc3339("2011-11-04T00:05:23.283+04:00")
d.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
# '2011-11-04T00:05:23.283000Z' <- Not actually the correct timezone
d.isoformat()
# '2011-11-04T00:05:23.283000+04:00'
d.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
# '2011-11-03T20:05:23.283000Z'
d.astimezone(timezone.utc).isoformat()
# '2011-11-03T20:05:23.283000+00:00'
TypeAdapter(datetime).dump_json(d)
# b'"2011-11-04T00:05:23.283000+04:00"'
TypeAdapter(datetime).dump_json(d.astimezone(timezone.utc))
# b'"2011-11-03T20:05:23.283000Z"'
```
All of those are valid RFC3339 outputs, but one of them is lying about the timezone. 

And now looking back at https://github.com/radiantearth/stac-spec/blob/v1.0.0/item-spec/item-spec.md#properties-object seems like there is actually a bit more of an issue. The spec requires UTC on top of the RFC3339. So even with `parse_rfc3339` that part is not being enforced on the input. So it would depend on if that needs to be enforced on input or not. 

I see a few possibilities:
```python
def parse_rfc3339_asutc(value: Any) -> datetime:
    return parse_rfc3339(value).astimezone(timezone.utc)

def parse_utc_rfc3339(value: Any) -> datetime:
    d = parse_rfc3339(value)
    if not d.tzinfo == timezone.utc:
        raise ValueError("Input must be UTC")
    return d

# Flexible input as long as there is a timezone. Converts to UTC. Use isoformat to get valid RFC format. 
Annotated[AwareDatetime, AfterValidator(lambda d: d.astimezone(timezone.utc)), PlainSerializer(lambda d: d.isoformat())]
# Input must be RFC, any timzone. Converts to UTC. Pydantic serializer outputs valid RFC format.
Annotated[datetime, BeforeValidator(parse_rfc3339_asutc)]
# Input must be RFC and UTC. Pydantic serializer outputs valid RFC format.
Annotated[datetime, BeforeValidator(parse_utc_rfc3339)]
```

I would probably still go with Pydantic for parsing. Gives the most flexibility on input, and can still guarantee valid output.

Reply:
I agree with @gadomski comments below in regards to "be permissive for inputs, strict on outputs."
[Pystac solution](https://github.com/stac-utils/pystac/blob/main/pystac/utils.py#L370C51-L409) for this uses `dateutil.parser` for converting strings and setting datetimes with missing timezone to UTC  when parsing to json.

This has the advantage that any date format can be accepted including `2020`, `2020-01` and `2020-01-01`.

I [added a possible solution](https://github.com/stac-utils/stac-pydantic/pull/135/files#diff-67202a6d8ddb30a3331bd88e34aefb079d099584598cf29afc95a55a1c5c8a54R44-R50) in here

Reply:
@thomas-maschler if going the `dateutil.parser` route, it would probably make sense to just let pydantic do its thing instead. It should be more performant. 

The way it is current written in this PR is to accept a datetime with a timezone, and convert it to UTC. Rather than accept a naive value and assume UTC. Though the spec here does say it must be in UTC. So, its not a horrible thing to assume either. Tons of possible solutions. I was trying to optimize on keeping as much as possible inside the pydantic rust code and minimal python code to maintain here. 

Its funny how a spec trying to simplify things or add clarity can make things more complicated than necessary.